src/render/vertex_array_object.js

```diff
@@ -128,42 +115,26 @@ class VertexArrayObject {
         }
 
         layoutVertexBuffer.enableAttributes(gl, program);
-        for (const vertexBuffer of paintVertexBuffers) {
-            vertexBuffer.enableAttributes(gl, program);
-        }
-
-        if (dynamicVertexBuffer) {
-            dynamicVertexBuffer.enableAttributes(gl, program);
-        }
-        if (dynamicVertexBuffer2) {
-            dynamicVertexBuffer2.enableAttributes(gl, program);
-        }
-        if (dynamicVertexBuffer3) {
-            dynamicVertexBuffer3.enableAttributes(gl, program);
-        }
-
         layoutVertexBuffer.bind();
         layoutVertexBuffer.setVertexAttribPointers(gl, program, vertexOffset);
+
         for (const vertexBuffer of paintVertexBuffers) {
+            vertexBuffer.enableAttributes(gl, program);
             vertexBuffer.bind();
             vertexBuffer.setVertexAttribPointers(gl, program, vertexOffset);
         }
 
-        if (dynamicVertexBuffer) {
-            dynamicVertexBuffer.bind();
-            dynamicVertexBuffer.setVertexAttribPointers(gl, program, vertexOffset);
+        for (const dynamicBuffer of dynamicVertexBuffers) {
+            if (dynamicBuffer) {
```

Comment:
Yes they can as some variant of the code may add some of these extensions buffers in some situations (such as with globe, when line-gradient are used, etc...). Checking for nullity at this level reduces the code needed from the callee side. Before, this was implicit with the type as it was already optional. Otherwise, we'd fully expand nullity checks for each of the data we add in this array:

```js
const dynamicBuffers = [];
if (painter.terrain && bucket.centroidVertexBuffer) dynamicBuffers.push(bucket.centroidVertexBuffer);
if (isGlobeProjection && bucket.layoutVertexExtBuffer) dynamicBuffers.push(bucket.layoutVertexExtBuffer);

program.draw(..., dynamicBuffers);
```
instead of
```js
const dynamicBuffers = [];
if (painter.terrain) dynamicBuffers.push(bucket.centroidVertexBuffer);
if (isGlobeProjection) dynamicBuffers.push(bucket.layoutVertexExtBuffer);

program.draw(..., dynamicBuffers);
```

or:
```js
program.draw(..., bucket.layoutVertexBuffer2 ? [bucket.layoutVertexBuffer2] : []);
```
instead of:
```js
program.draw(..., [bucket.layoutVertexBuffer2]);
```